### DISCUSSION TITLE
CCSDS Reed solomon and ASM marker

### DISCUSSION THREAD

**Discussion by keenerm** [daniestevez/gr-satellites#553] [Q&A] [answered]
Hi,
I am creating a CCSDS concatenated encoder with Reed Solomon and a convolutional encoder.
I am trying to utilize satellites Reed Solomon encoder for RS(255/223) and RS(255/239) from what I understand the Generator polynomial:285 is a primitive polynomial that generates the elements for a GF(2^8) field. So, for my cases do I just feed an input frame of either 239- or 223-bytes length? also does the encoder block automatically create and append the parity blocks?

For the ASM marker part I thought of using the "sync and create pdu block" (as shown here https://github.com/daniestevez/gr-satellites/discussions/461 ) after my Reed Solomon encoder(not sure if i did so correctly). But I'm wondering what to put after the Viterbi block to read the marker and sync the data correctly before going to the Reed Solomon decoder.  

Any Help is appreciated thanks!

<img width="1168" alt="GNU" src="https://github.com/daniestevez/gr-satellites/assets/172632607/7dba1806-5f75-45a7-85d6-552e877a54d5">

**Comment by daniestevez** [comment, marked as answer]
Hello, some comments:
- The "CCSDS Reed-Solomon Encoder" block supports the (255,223) CCSDS code, both with the dual and the conventional basis, but it doesn't support the (255, 239) code.
- You could configure the generic "Reed-Solomon Encoder" block to use the (255, 239) code with the conventional basis, but I don't think there's out-of-the box support for the (255, 239) with the dual basis. Some C++ coding will be needed.
- "Sync and Create PDU" doesn't serve to _add_ an ASM. It serves to _detect_ an ASM and extract a PDU with the frame following the ASM. Hence it belongs to a receiver. Not to a transmitter.
- For the receiver you already have a fully working solution in the "CCSDS Concatenated Deframer" block, except that it doesn't support the (255, 239) code.

> So, for my cases do I just feed an input frame of either 239- or 223-bytes length?

Yes. The input to any of the Reed-Solomon encoder blocks should be a PDU with the information work packed as 8 bits per byte.

> also does the encoder block automatically create and append the parity blocks?

This question puzzles me. That's precisely what the Reed-Solomon Encoder does. What else could it do if it didn't do this?

**Reply by keenerm** [reply]
I have added the ASM marker and retrieved my transmitted data. I have tried both the CCSDS RS encode/decode pair and the generic RS encode/decode pair. The CCSDS pair works well for the (255,223) case and outputs the correct 255-byte/symbol codeword size, the generic however only outputs a codeword that is +1 of what information length is input. For example, if 223 data bytes (symbols) were sent to it, it will only output 224 symbol codeword or for 239 symbols it will output a 240 codeword. 
I wanted to ask what must be input to get the 255 symbol codeword outputs.
Also, I wanted to ask where I can find the source code for the Generic Rs blocks so that I may potentially adjust the generic block for (255,239) dual basis.

![image](https://github.com/daniestevez/gr-satellites/assets/172632607/81a9bae1-d6b5-4b9b-b94b-62001d87d406)

**Reply by daniestevez** [reply]
> the generic however only outputs a codeword that is +1 of what information length is input.

This is determined by the number of roots parameter.

> where I can find the source code for the Generic Rs blocks

https://github.com/daniestevez/gr-satellites/blob/main/lib/decode_rs_impl.cc
https://github.com/daniestevez/gr-satellites/blob/main/lib/encode_rs_impl.cc

**Reply by keenerm** [reply]
Thank you, for all the help!
